Virtual envirnoment data dir issues when using local directory

apatsekin opened the pull request:
Hey guys! I believe there are few issues with how [VirtualEnvironment](https://github.com/pypa/hatch/blob/02eb8037fde7c7a9f59c332aa9ff091f3b65f748/src/hatch/env/virtual.py#L27) manages directories. 
Everything works fine until you try to set `HATCH_DATA_DIR=/root_project/.hatch/data` - i.e. to something inside the project.

`VirtualEnvironment.__init__()` checks if data_directory is inside root_directory and based on that either uses "flat structure" or adds a `checksum`-level of directories.:
```
 # Conditions requiring a flat structure
elif self.root in self.data_directory.parents or [...]:
    self.storage_path = self.data_directory
    self.virtual_env_path = self.storage_path / venv_name
# Otherwise the defined app path
else:
    self.storage_path = self.data_directory / project_name / checksum
    self.virtual_env_path = self.storage_path / venv_name
```
But since `self.data_directory` is not `.resolved()` it can either be absolute or relative path (both being inside self.root) - based on that `if` produces different results: `HATCH_DATA_DIR=/root_project/.hatch/data` - will get a "flat structure" while `HATCH_DATA_DIR=.hatch/data`- will get a directory with `checksum`. 

While `[env]-build` environment [will always go](https://github.com/pypa/hatch/blob/02eb8037fde7c7a9f59c332aa9ff091f3b65f748/src/hatch/env/virtual.py#L39) to `checksum` directory no matter what.

This behavior brings two issues:
1. When `HATCH_DATA_DIR=/root_project/.hatch/data` regular env ends up being in "flat-structure", while `-build` env goes to "checksum-structure". If user runs `hatch build` first, it will create env under: `root/.hatch/virtual/my-project/jeh852/my-project-build`. Then user runs `hatch run format` and hatch will fail, since it will [check](https://github.com/pypa/hatch/blob/02eb8037fde7c7a9f59c332aa9ff091f3b65f748/src/hatch/env/virtual.py#L104) for `default` env under flat structure: `root/.hatch/virtual/my-project/` - dir does exist, but there is no env there => hatch will crash on the next step.
2. While using `HATCH_DATA_DIR=.hatch/data` (relative path) - the flat structure is completely ignored, which doesn't cause errors, but ignores scenario described [here](https://hatch.pypa.io/latest/plugins/environment/virtual/#location).

~~nit: when user sets `envs.default.path=` property only regular environment is created under that path. I guess `-build` one also should be placed there, right? It's a resource of the same Environment.~~

**TL/DR**: proposing PR to properly handle "flat structure", when`HATCH_DATA_DIR` is inside root of the project.

apatsekin commented:
Hey @ofek, would be great if this one makes it to upcoming release.

ofek commented:
Yes it will be thank you!

> nit: when user sets envs.default.path= property only regular environment is created under that path. I guess -build one also should be placed there, right? It's a resource of the same Environment.

I want those to always be isolated

apatsekin commented:
> > nit: when user sets envs.default.path= property only regular environment is created under that path. I guess -build one also should be placed there, right? It's a resource of the same Environment.
> 
> I want those to always be isolated

Yeah they will be isolated, it's just that `envs.default.path=/my/path/to/env` will be respected by both build and non-build environments:
```
/my/path/to/env
/my/path/to/env-build
``` 
In current `master` it affects the non-build only, putting `-build` one to default path with default name:
```
/my/path/to/env
/default/path/.hatch/env/virtual/[project]/[hash]/[project]-build
```
which seems to be slightly confusing? If it's intentional, I'm happy to remove this part from PR.

ofek commented:
Yes actually I don't want build environments anywhere else but that internal hidden location away from sight

apatsekin commented:
Sounds good, I reverted original behavior for `envs.default.path`. Although `-build` environment respects `HATCH_DATA_DIR`, so when it's set to local project dir - all environments go there, right? This is really great feature, since we can configure hatch to create temp files only inside project dir without polluting global disk. But due to `-build` env ignoring local-dir check, those issues emerge described in the PR. So I changed it to minimal:
1. add a check for `-build` environment to use flat structure when necessary, based on `isolated_data_directory`
2. use `.resolve()` to check if `HATCH_DATA_DIR` is actually inside project root.
